lib/cn/cn_tree.c

```diff
@@ -1443,22 +1483,168 @@ cn_comp_update_spill(struct cn_compaction_work *work, struct kvset **kvsets)
+            le = list_last_entry(&pnode->tn_kvset_list, struct kvset_list_entry, le_link);
+            list_del(&le->le_link);
+            list_add(&le->le_link, &retired_kvsets);
+        }
+        assert(list_empty(&pnode->tn_kvset_list));
+
+        /* Repurpose the parent node as the right node and add kvsets to it.
+         * The parent node is already part of the cN tree and is protected by a compaction
+         * token.
+         */
+        right = pnode;
+        if (nodeidv[1] != CN_TREE_INVALID_NODEID) {
+            right->tn_nodeid = nodeidv[1];
+
+            INIT_LIST_HEAD(&right->tn_kvset_list);
+            for (k = w->cw_kvset_cnt; k < w->cw_outc; k++) {
+                if (kvsets[k])
+                    kvset_list_add_tail(kvsets[k], &right->tn_kvset_list);
+            }
+
+            /* The last node in the route map contains all keys that are greater than the
+             * penultimate node. Under rare circumstances the split key chosen for the last
+             * node can be lexicographically greater than its edge key. The below logic
+             * detects this situation and updates the edge key of the right node to its
+             * max key at the time of split.
+             */
+            if (route_node_islast(right->tn_route_node)) {
+                int rc = route_node_keycmp(right->tn_route_node, w->cw_split.key, w->cw_split.klen);
+                if (ev(rc <= 0)) {
+                    err = route_node_key_modify(tree->ct_route_map, right->tn_route_node,
+                                                rekey, reklen);
+                    if (err) {
+                        rmlock_wunlock(&tree->ct_lock);
+                        goto exit;
+                    }
+                }
+            }
+            assert(route_node_keycmp(right->tn_route_node, w->cw_split.key, w->cw_split.klen) > 0);
+
+            w->cw_split.nodev[1] = right;
+        }
+
+        /* Update route map with the left edge and add the new left node to the cN tree list.
+         * The right node is already part of the cn tree list.
+         */
+        if (left) {
+            left->tn_route_node = route_map_insert(tree->ct_route_map, left,
+                                                   w->cw_split.key, w->cw_split.klen);
+            if (!left->tn_route_node) {
+                rmlock_wunlock(&tree->ct_lock);
+                err = merr(ENOMEM);
+                goto exit;
+            }
+
+            list_add_tail(&left->tn_link, &tree->ct_nodes);
+        }
+
+        /* Update samp stats
+         */
+        for (int i = 0; i < 2 && w->cw_split.nodev[i]; i++) {
+            cn_tree_samp(tree, &w->cw_samp_pre);
+            cn_tree_samp_update_compact(tree, w->cw_split.nodev[i]);
+            cn_tree_samp(tree, &w->cw_samp_post);
+        }
+
+        atomic_sub_rel(&pnode->tn_busycnt, (1u << 16) + w->cw_kvset_cnt);
+    }
+    rmlock_wunlock(&tree->ct_lock);
```

Comment:
You could use a do-while loop so that you can use break rather than have to drop the lock and use goto, e.g.:

```
wlock(ct_lock);
do {
    stuff;
    if (err)
        break;
} while (0);
wunlock(ct_lock);

if (err) {
    cn_node_free();
    return err;
}

listf_foreach...() {
}

return 0;
```

Reply:
Nice suggestion, thanks!